### PR TITLE
Fixing Path Params

### LINKED ISSUE #17: ArrayIndexOutOfBoundsException thrown when (optional or required)path param not passed
The Buildable is detected and when the last path param is not found and exception is thrown with 500 server error.
Expected behaviour, 404 not found should be returned

### PR DESCRIPTION
**Pull request by shrey-garg**
Fixes #17

### CODE REVIEW

File: api/src/main/java/com/flipkart/poseidon/api/APILegoSet.java [hunk 1]
```diff
@@ -65,10 +65,7 @@ public Buildable getBuildable(Request request) throws LegoSetException, ElementN
         String completeUrl = ApiHelper.getUrlWithHttpMethod(poseidonRequest.getUrl(), httpMethod);
         Buildable buildable = trie.get(getKeysForTrie(completeUrl));
         if (buildable == null) {
-            buildable = trie.get(getKeysForTrie(poseidonRequest.getUrl()));
-            if (buildable == null) {
-                throw new ElementNotFoundException("Buildable not found for given url: " + poseidonRequest.getUrl());
-            }
+            throw new ElementNotFoundException("Buildable not found for given url: " + poseidonRequest.getUrl());
```
Comment (pmohankumar) [inline api/src/main/java/com/flipkart/poseidon/api/APILegoSet.java]: With this change, httpMethod is now mandatory in API definition. Please update https://github.com/flipkart-incubator/Poseidon/wiki/Migration-Guide-%5B3.x-to-4.x%5D

File: api/src/main/java/com/flipkart/poseidon/api/APILoader.java [hunk 2]
```diff
@@ -67,7 +68,7 @@ public APILoader(PoseidonLegoSet legoSet, String config, Configuration configura
                         completeUrl = ApiHelper.getUrlWithHttpMethod(pojo.getUrl(), pojo.getHttpMethod().toString());
                     } else {
                         // to remove extra slashes in url
-                        completeUrl = ApiHelper.getUrlWithHttpMethod(pojo.getUrl(), null);
+                        completeUrl = ApiHelper.getUrlWithHttpMethod(pojo.getUrl(), HttpMethod.GET.toString());
```
Comment (pmohankumar) [inline api/src/main/java/com/flipkart/poseidon/api/APILoader.java]: If httpMethod is null, throw an error which gets caught and logged.

File: api/src/main/java/com/flipkart/poseidon/api/APILegoSet.java (hunk 1 above)
Comment (shrey-garg) [inline api/src/main/java/com/flipkart/poseidon/api/APILegoSet.java]: Done

File: api/src/main/java/com/flipkart/poseidon/internal/ParamValidationFilter.java [hunk 3]
```diff
@@ -106,7 +106,7 @@ else if(isBodyRequest) {
                 } else if (isPathParam) {
                     int pos = param.getPosition();
                     String[] splitUrl = poseidonRequest.getUrl().split("/");
-                    value = splitUrl[pos];
+                    value = splitUrl.length > pos ? splitUrl[pos] : null;
```
Comment (pmohankumar) [inline api/src/main/java/com/flipkart/poseidon/internal/ParamValidationFilter.java]: Make it a if () { value = } else throw exception.

**Pull request merged**